libfreerdp/core/gcc.c

```diff
@@ -829,49 +853,139 @@ BOOL gcc_write_user_data_header(wStream* s, UINT16 type, UINT16 length)
 	return TRUE;
 }
 
-static UINT16 earlyCapsFromSettings(const rdpSettings* settings)
+static UINT32 filterAndLogEarlyServerCapabilityFlags(UINT32 flags)
+{
+	const UINT32 mask =
+	    (RNS_UD_SC_EDGE_ACTIONS_SUPPORTED_V1 | RNS_UD_SC_DYNAMIC_DST_SUPPORTED |
+	     RNS_UD_SC_EDGE_ACTIONS_SUPPORTED_V2); // TODO: Not implemented
+	                                           // RNS_UD_SC_SKIP_CHANNELJOIN_SUPPORTED
+	const UINT32 filtered = flags & mask;
+	const UINT32 unknown = flags & (~mask);
+	if (unknown != 0)
+	{
+		char buffer[256] = { 0 };
+		WLog_WARN(TAG,
+		          "TS_UD_SC_CORE::EarlyCapabilityFlags [0x%08" PRIx32 " & 0x%08" PRIx32
+		          " --> 0x%08" PRIx32 "] filtering %s, feature not implemented",
+		          flags, ~mask, unknown,
+		          rdp_early_server_caps_string(unknown, buffer, sizeof(buffer)));
+	}
+	return filtered;
+}
+
+static UINT32 earlyServerCapsFromSettings(const rdpSettings* settings)
 {
-	UINT16 earlyCapabilityFlags = settings->EarlyCapabilityFlags;
+	UINT32 EarlyCapabilityFlags = 0;
+
+	if (settings->SupportEdgeActionV1)
+		EarlyCapabilityFlags |= RNS_UD_SC_EDGE_ACTIONS_SUPPORTED_V1;
+	if (settings->SupportDynamicTimeZone)
+		EarlyCapabilityFlags |= RNS_UD_SC_DYNAMIC_DST_SUPPORTED;
+	if (settings->SupportEdgeActionV2)
+		EarlyCapabilityFlags |= RNS_UD_SC_EDGE_ACTIONS_SUPPORTED_V2;
+	if (settings->SupportSkipChannelJoin)
+		EarlyCapabilityFlags |= RNS_UD_SC_SKIP_CHANNELJOIN_SUPPORTED;
+
+	return filterAndLogEarlyServerCapabilityFlags(EarlyCapabilityFlags);
+}
+
+static UINT16 filterAndLogEarlyClientCapabilityFlags(UINT32 flags)
+{
+	const UINT32 mask =
+	    (RNS_UD_CS_SUPPORT_ERRINFO_PDU | RNS_UD_CS_WANT_32BPP_SESSION |
+	     RNS_UD_CS_SUPPORT_STATUSINFO_PDU | RNS_UD_CS_STRONG_ASYMMETRIC_KEYS |
+	     RNS_UD_CS_VALID_CONNECTION_TYPE | RNS_UD_CS_SUPPORT_MONITOR_LAYOUT_PDU |
+	     RNS_UD_CS_SUPPORT_NETCHAR_AUTODETECT | RNS_UD_CS_SUPPORT_DYNVC_GFX_PROTOCOL |
+	     RNS_UD_CS_SUPPORT_DYNAMIC_TIME_ZONE |
+	     RNS_UD_CS_SUPPORT_HEARTBEAT_PDU); // unsupported RNS_UD_CS_SUPPORT_SKIP_CHANNELJOIN
+	const UINT32 filtered = flags & mask;
+	const UINT32 unknown = flags & ~mask;
+	if (unknown != 0)
+	{
+		char buffer[256] = { 0 };
+		WLog_WARN(TAG,
+		          "(TS_UD_CS_CORE)::EarlyCapabilityFlags [0x%08" PRIx32 " & 0x%08" PRIx32
+		          " --> 0x%08" PRIx32 "] filtering %s, feature not implemented",
+		          flags, ~mask, unknown,
+		          rdp_early_client_caps_string(unknown, buffer, sizeof(buffer)));
```

Comment:
Don't think it makes sense to call `rdp_early_client_caps_string` since we already know it is an unknown flag and we already log all the flags in this message

Reply:
there is the `RNS_UD_CS_SUPPORT_SKIP_CHANNELJOIN` flag that is not implemented yet.